src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
+        encryptionMethod: this.NEW_USER_ENCRYPTION_METHOD,
+        password,
+      },
+      quota: undefined,
+      isOverQuota: false
     };
-    this.keyConfig.data().accessKeys.push(accessKeyJson);
-    try {
-      this.keyConfig.write();
-    } catch (error) {
-      throw new Error(`Failed to save config: ${error}`);
-    }
+    this.accessKeys.push(accessKey);
+    this.saveAccessKeys();
     await this.updateServer();
-    return makeAccessKey(this.proxyHostname, accessKeyJson);
+    return accessKey;
   }
 
   removeAccessKey(id: AccessKeyId): boolean {
-    for (let ai = 0; ai < this.keyConfig.data().accessKeys.length; ai++) {
-      const accessKey = this.keyConfig.data().accessKeys[ai];
+    for (let ai = 0; ai < this.accessKeys.length; ai++) {
+      const accessKey = this.accessKeys[ai];
       if (accessKey.id === id) {
-        this.portProvider.freePort(accessKey.port);
-        this.keyConfig.data().accessKeys.splice(ai, 1);
-        this.keyConfig.write();
+        this.portProvider.freePort(accessKey.proxyParams.portNumber);
+        this.accessKeys.splice(ai, 1);
+        this.saveAccessKeys();
         this.updateServer();
         return true;
       }
     }
     return false;
   }
 
-  listAccessKeys(): IterableIterator<AccessKey> {
-    return this.keyConfig.data().accessKeys.map(
-        accessKeyJson => makeAccessKey(this.proxyHostname, accessKeyJson))[Symbol.iterator]();
+  listAccessKeys(): AccessKey[] {
+    // Deep copy the access key array to prevent returning references to keys or their properties.
+    return this.accessKeys.map(key => (key instanceof Object) ? {...key} : key);
   }
 
   renameAccessKey(id: AccessKeyId, name: string): boolean {
-    const accessKeyJson = this.getAccessKey(id);
-    if (!accessKeyJson) {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
       return false;
     }
-    accessKeyJson.name = name;
+    accessKey.name = name;
     try {
-      this.keyConfig.write();
+      this.saveAccessKeys();
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async setAccessKeyQuota(id: AccessKeyId, quota: AccessKeyQuota): Promise<boolean> {
+    if (!quota || quota.quotaBytes < 1 || quota.windowHours < 1) {
+      return false;
+    }
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    accessKey.quota = quota;
+    try {
+      this.saveAccessKeys();
+      const quotaStausChanged = await this.updateAccessKeyQuotaStaus(accessKey);
+      if (quotaStausChanged) {
+        // Reflect the access key quota status if it changed with the new quota.
+        await this.updateServer();
+      }
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async removeAccessKeyQuota(id: AccessKeyId): Promise<boolean> {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    const quotaStausChanged = accessKey.isOverQuota;
```

Comment:
This variable name seems off to me.

Reply:
Renamed to `wasOverQuota` in 2105d3f, which is more specific.